packages/inter-protocol/src/vaultFactory/vault.js

```diff
@@ -480,77 +514,144 @@ export const prepareVault = (baggage, marshaller, zcf) => {
-            proposal.give.Minted || helper.emptyDebt(),
-            debt,
-          );
-          const wantMinted = proposal.want.Minted || helper.emptyDebt();
-          if (allEmpty([giveColl, giveMinted, wantColl, wantMinted])) {
-            clientSeat.exit();
-            return 'no transaction, as requested';
-          }
-
-          // Calculate the fee, the amount to mint and the resulting debt. We'll
-          // verify that the target debt doesn't violate the collateralization ratio,
-          // then mint, reallocate, and burn.
-          const { newDebt, fee, toMint } = helper.loanFee(
-            debt,
-            giveMinted,
-            wantMinted,
-          );
-
-          trace('adjustBalancesHook', state.idInManager, {
+          trace('adjustBalancesHook after quote', state.idInManager, {
             newCollateralPre,
-            newCollateral,
             fee,
             toMint,
+            maxDebtPre,
             newDebt,
           });
 
+          // While awaiting maxDebtFor, the allocations may have changed.
+          // After the `await`, we retrieve the vault's allocations again,
+          // so we can compare to the debt limit based on the new values.
+          const collateral = helper.getCollateralAllocated(vaultSeat);
+          const newCollateral = addSubtract(
+            collateral,
+            fp.give.Collateral,
+            fp.want.Collateral,
+          );
           if (
-            checkRestart(newCollateralPre, maxDebtPre, newCollateral, newDebt)
+            allocationsChangedSinceQuote(
+              newCollateralPre,
+              maxDebtPre,
+              newCollateral,
+              newDebt,
+            )
           ) {
             return helper.adjustBalancesHook(clientSeat);
           }
 
-          stageDelta(clientSeat, vaultSeat, giveColl, wantColl, 'Collateral');
+          return helper.commitBalanceAdjustment(
+            clientSeat,
+            fp,
+            {
+              newDebt,
+              fee,
+              surplus,
+              toMint,
+            },
+            { normalizedDebtPre, collateralPre },
+          );
+        },
+
+        /**
+         *
+         * @param {ZCFSeat} clientSeat
+         * @param {FullProposal} fp
+         * @param {ReturnType<typeof calculateLoanCosts>} costs
+         * @param {object} accounting
+         * @param {NormalizedDebt} accounting.normalizedDebtPre
+         * @param {Amount<'nat'>} accounting.collateralPre
+         * @returns {Promise<string>} success message
+         */
+        async commitBalanceAdjustment(
+          clientSeat,
+          fp,
+          { newDebt, fee, surplus, toMint },
+          { normalizedDebtPre, collateralPre },
+        ) {
+          const { state, facets } = this;
+          const { helper } = facets;
+          const { vaultSeat } = state;
+
+          const takeMinted = AmountMath.subtract(fp.give.Minted, surplus);
+
+          stageDelta(
```

Comment:
stageDelta is deprecated. Would you mind switching to atomicRearrange?

Reply:
I had tried and it's non-trivial given that `state.manager.mintAndReallocate` is doing the reallocation.

> I'd like to see a switch to atomicRearrange. If that's too much for this PR, how soon?

I'd say so. I expect it to come with https://github.com/Agoric/agoric-sdk/issues/5605